FWCore/Framework/src/Principal.cc

```diff
@@ -171,9 +171,20 @@ namespace edm {
     if(hasAliases) {
       for(auto const& prod : prodsList) {
         BranchDescription const& bd = prod.second;
-        if(bd.isAlias() && bd.branchType() == branchType_) {
+        if(bd.isAnyAlias() && bd.branchType() == branchType_) {
           auto cbd = std::make_shared<BranchDescription const>(bd);
-          addAliasedProduct(cbd);
+          if(bd.isSwitchAlias()) {
+            if(bd.onDemand()) {
```

Comment:
Maybe some documentation here explaining the reason?

Reply:
For the differentiation of on-demand and non-on-demand `SwitchProducer`? Will add.